### DISCUSSION TITLE
Two (or more) communication interfaces for one device

### DISCUSSION THREAD

**Discussion by robertkowalewski** [zephyrproject-rtos/zephyr#64108] [Q&A]
Hi,
I'm writing my own driver to zephyr for BMS device, and it has a possibility to work in dual access ring topology. In my case there may be two SPIs... And this makes my problem. There wouldn't be a problem if I allow only to use one MCU SPI but with different CS pins and connect them respectively to both interfaces in my BMS. But I'd like to have more flexibility in my driver and allows to use two different MCU SPIs for one specific device? Is it possible to achieve? If so, then please let me know how I can achieve that.

I hope that my question is pretty clear.

**Comment by kartben** [comment]
Could this https://github.com/zephyrproject-rtos/zephyr/pull/62998 help?

**Reply by robertkowalewski** [reply]
@kartben - actually it looks like very sensible proposal, but... it doesn't meet my expectation. I'd like to implement a mechanism of choosing which SPI interface should be chosen within a device driver. In case of #62998, as far as I understand that change, the decision which device should be used is left to application, am I right?

Maybe I'll present the whole picture of my case and it'll allow to find the proper solution. The one instance of my driver covers all devices which are connected in dual ring topology.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/15970910/67379d51-2aae-40f8-b773-499d029606ba)

In typical conditions MCU will communicate with devices via SPI1 (each device may communicate with its neighbor and forwards message sent to them by MCU), but when communication fails on SPI1, I'd like to try (driver would be responsible for that) to send SPI messages via SPI2. 

So what is recommended way to achieve that?

**Comment by ndrs-pst** [comment]
@robertkowalewski, looking at your diagram, it seems like implementing the BMS drivers out-of-tree and passing two SPI devices for internal use might be the way to go.

**Reply by robertkowalewski** [reply]
@ndrs-pst - thanks for answer. Do you know any example of such a solution in zephyr tree?

**Reply by ndrs-pst** [reply]
@robertkowalewski I'm not sure if there's such an example in the Zephyr tree, but this subsystem looks promising to me: https://docs.zephyrproject.org/3.5.0/services/sensing/index.html
